text/0000-build-plan-merging.md

```diff
@@ -0,0 +1,260 @@
+- Putting this information in the `requires.metadata` obfuscates the flow of information between buildpacks.
+- This implicit contract between dependent buildpacks should be formalized.
+
+### Example:
+ We consider the interactions between the Paketo `node-engine` and `npm`:
+
+ - The `node-engine` provides the `node` dependency to an image, but only for the `launch` phase.
+ - The `npm` requires `npm` during the `build` phase, and so it adds an entry in `requires.metadata` to reflect this need, it may also require a version of `node-engine` so it may `require` a version of `node-engine`.
+ - As a result during the `node-engine`'s build phase all `entries.metadata` must be parsed for `build`, `launch`, and `version` to determine the correct flags to attach to the `npm` dependency Layer Metadata.
+
+## Goal
+1. Simplify the Build Plan (TOML) and Buildpack Plan (TOML) formats.
+2. Clarify and codify how buildpacks communicate with each other.
+3. Leverage the lifecycle to implement task that would simplify the buildpack authoring process.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Build Plan (TOML) Changes
+We propose a complete restructuring of the `requires` field and an addition to the `provides` field in the Build Plan (TOML) file.
+
+We propose the following Build Plan (TOML):
+
+```toml
+[[provides]]
+name = "<dependency name>"
+[provides.strategy.version]
+  collect = false
+
+[[requires]]
+name = "<dependency name>"
+[requires.version]
+  constraint = "<dependency constraint>"
+  source = "<dependency constraint source>"
+[requires.capabilities]
+  # arbitary keys to bools
+  build = false
+  launch = false
+[requires.modifiers]
+  # arbitrary keys and values
+  some-modifer = "modifier"
+
+[[or]]
+
+[[or.provides]]
+name = "<dependency name>"
+[provides.strategy.version]
+  collect = false
+
+[[or.requires]]
+name = "<dependency name>"
+[requires.version]
+  constraint = "<other dependency constraint>"
+  source = "<other dependency constraint source>"
+[requires.capabilities]
+  # arbitary keys to bools
+  build = false
+  launch = false
+[requires.modifiers]
+  # arbitrary keys and values
+  some-other-modifer = "other modifier"
+```
+
+This proposal adds a new field to `provides`:
+- `provides.strategy.version` field holds a `collect` key which indicates if the version fields from all of the requires should be merged together or if they should collected into an array.
+  - This field is structured to be extensible to other fields in `requires`.
+
+This proposal also removes both the `requires.version` and `requires.metadata` fields and replaces them with three new fields each with their own purpose:
+- `requires.version` field holds all of the version requirement information.
+  - `constraint` can be any semver constraint used to decide which dependency the providing buildpack will install.
+  - `source` is the origin of the constraint (i.e. it came from `package.json`).
```

Comment:
Is this field required? Also, what do you use it for, the BOM at the end?

Reply:
+1 to @hone 's question